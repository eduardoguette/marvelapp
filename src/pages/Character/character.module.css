.characterHero {
  background-color: var(--secondary);
  border-top: 1px solid var(--gray);
  color: var(--white);

}

.characterCard {
  margin: auto;
  display: flex;
  justify-content: center;
}

.characterImage {
  aspect-ratio: 1;
  max-width: 320px;
}

.characterInfo {
  max-width: 640px;
  flex: auto;
  display: flex;
  flex-direction: column;
  justify-content: center;
  padding: 3rem;
}

.characterTitle {
  font: var(--headline2);
  text-transform: uppercase;
}

.characterContainer {
  padding: 48px 0px 48px 0px;
}
.characterCommics {
  padding: 0px 3rem 0px 3rem;
  display: flex;
  flex-direction: column;
  gap: 1.5rem;
}