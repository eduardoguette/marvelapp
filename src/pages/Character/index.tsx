import { characters } from '@/api'
import { CommicsCard, Skeleton } from '@/components'
import Carrousel from '@/components/Carrousel'
import { Action } from '@/components/Characters'
import { Loader } from '@/components/commons/Loader'
import { Characters } from '@/types'
import { CommicsData } from '@/types/CharactersCommics'
import { useQuery } from '@tanstack/react-query'
import { AxiosResponse } from 'axios'
import { AnimatePresence, motion } from 'framer-motion'
import { useParams } from 'react-router-dom'
import styles from './character.module.css'

export const Character = () => {
  const { id } = useParams()
  const { isLoading, data, isError } = useQuery<AxiosResponse<Characters>>({
    queryFn: () => characters.getById(id as string),
    queryKey: ['character', id],
    staleTime: Infinity,
  })
  const { data: dataCommics, isLoading: isLoadingCommics } = useQuery<
    AxiosResponse<CommicsData>
  >({
    queryFn: () => characters.getCommics(id as string),
    queryKey: ['character/commics', id],
    staleTime: Infinity,
  })

  if (isLoading || isLoadingCommics)
    return (
      <>
        <Skeleton
          style={{
            width: '100%',
            height: '320px',
          }}
        />
      </>
    )

  if (!data) return <div>Not found</div>
  if (isError) return <div>Error</div>

  const character = data.data?.data?.results?.[0]

  if (!character) return <div>No character found</div>

  return (
    <div>
      <section className={styles.characterHero}>
        <motion.article
          initial={{
            opacity: 0,
            x: -50,
          }}
          animate={{
            opacity: 1,
            x: 0,
          }}
          exit={{
            opacity: 0,
            x: 50,
          }}
          className={styles.characterCard}
        >
          <img
            className={styles.characterImage}
            src={`${character.thumbnail.path}.${character.thumbnail.extension}`}
            alt={character.name}
          />
          <div className={styles.characterInfo}>
            <header className={styles.characterCardHeader}>
              <motion.h2
                initial={{
                  opacity: 0,
                }}
                animate={{
                  opacity: 1,
                }}
                className={styles.characterTitle}
              >
                {character.name}
              </motion.h2>
              <Action idCharacter={Number(id)} characterDetails />
            </header>
            <motion.p
              initial={{
                opacity: 0,
              }}
              animate={{
                opacity: 1,
              }}
            >
              {character.description}
            </motion.p>
          </div>
        </motion.article>
      </section>
      {dataCommics && (
        <section className={styles.characterContainer}>
          <article className={styles.characterCommics}>
            <h2 className={styles.characterTitle}>Commics</h2>
            {dataCommics?.data.data.results.length > 0 && (
              <AnimatePresence>
                <Carrousel>
                  {dataCommics?.data.data.results.map((commic, index) => (
                    <motion.div
                      key={index}
                      initial={{ opacity: 0 }}
                      animate={{ opacity: 1 }}
                      exit={{ opacity: 0 }}
                      transition={{ duration: 0.5, delay: index * 0.3 }}
                    >
                      <Carrousel.Item>
                        <CommicsCard commic={commic} />
                      </Carrousel.Item>
                    </motion.div>
                  ))}
                </Carrousel>
              </AnimatePresence>
            )}
            {!isLoadingCommics &&
              dataCommics?.data.data.results.length === 0 && (
                <p>No commics found</p>
              )}
            {isLoadingCommics && <Loader />}
          </article>
        </section>
      )}
    </div>
  )
}
