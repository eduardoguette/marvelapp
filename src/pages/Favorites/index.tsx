import { characters } from '@/api'
import { CharactersList } from '@/components/Characters'
import { Filter } from '@/components/Characters/Filter'
import { ContextAppProvider } from '@/context'

import { Loader } from '@/components/commons/Loader'
import { Characters } from '@/types'
import { useQuery } from '@tanstack/react-query'
import { AxiosResponse } from 'axios'
import { useContext, useEffect } from 'react'
import styles from './favorites.module.css'
export const Favorites = () => {
  const { setData, filter, favorites } = useContext(ContextAppProvider)

  const { isLoading, data: dataResult } = useQuery<AxiosResponse<Characters>>({
    queryFn: characters.getAll,
    queryKey: ['characters']
  })

  useEffect(() => {
    if (dataResult)
      setData(
        dataResult.data.data.results
          .filter((item) =>
            favorites.find((itemCharacter) => itemCharacter === item.id)
          )
          .filter((item) =>
            item.name.toLowerCase().includes(filter.toLowerCase())
          )
      )
  }, [dataResult, filter, setData, favorites])

  if (isLoading) return <Loader />
  return (
    <section className={styles.container}>
      <h2 className={styles.h2}>Favorites</h2>
      <Filter />
      <CharactersList />
    </section>
  )
}
