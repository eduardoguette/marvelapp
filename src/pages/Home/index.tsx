import { characters } from '@/api'
import { CharactersList } from '@/components/Characters'
import { Filter } from '@/components/Characters/Filter'
import { ContextAppProvider } from '@/context'
import { Characters } from '@/types'
import { useQuery } from '@tanstack/react-query'
import { AxiosResponse } from 'axios'
import { motion } from 'framer-motion'
import { useContext, useEffect } from 'react'
import styles from './home.module.css'

export const Home: React.FC = () => {
  const { setData, filter } = useContext(ContextAppProvider)

  const { isLoading, data } = useQuery<AxiosResponse<Characters>>({
    queryFn: characters.getAll,
    queryKey: ['characters']
  })

  useEffect(() => {
    if (!isLoading && data) {
      setData(
        data.data.data.results.filter((item) =>
          item.name.toLowerCase().includes(filter.toLowerCase())
        )
      )
    }
  }, [data, filter, setData, isLoading])

  if (isLoading) return <div>Loading...</div>

  return (
    <motion.div
      initial={{
        opacity: 0,
        x: 100
      }}
      animate={{
        opacity: 1,
        x: 0
      }}
      exit={{
        opacity: 0,
        x: -100
      }}
      className={styles.container}
    >
      <Filter />
      <CharactersList />
    </motion.div>
  )
}
