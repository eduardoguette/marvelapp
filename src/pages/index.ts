export * from './Character'
export * from './Favorites'
export * from './Home'
