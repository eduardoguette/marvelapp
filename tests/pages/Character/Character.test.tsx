import { useQuery } from '@tanstack/react-query'
import { render, screen } from '@testing-library/react'
import React from 'react'
import { useParams } from 'react-router-dom'
import { Mock, describe, expect, it, vi } from 'vitest'
import { Character } from '../../../src/pages/Character'

const mockCharacterData = {
  data: {
    data: {
      results: [
        {
          id: '1',
          name: 'John Doe',
          description: 'Test description',
          thumbnail: {
            path: 'https://example.com/image',
            extension: 'jpg'
          }
        }
      ]
    }
  }
}
const mockCommicsData = {
  data: {
    data: {
      results: [
        {
          id: '1',
          title: 'Amazing Spider-Man',
          thumbnail: {
            path: 'https://example.com/commic-image',
            extension: 'jpg'
          },
          modified: '2021-04-2'
        }
      ]
    }
  }
}

vi.mock('react-router-dom', () => ({
  useParams: vi.fn()
}))

vi.mock('@tanstack/react-query', () => ({
  useQuery: vi.fn()
}))

describe('Character Component', () => {
  const mockUseParams = useParams as Mock
  const mockUseQuery = useQuery as Mock

  it('Debe mostrar el Skeleton mientras estÃ¡ cargando', () => {
    mockUseParams.mockReturnValue({ id: '1' })
    mockUseQuery.mockReturnValue({
      isLoading: true,
      data: null,
      isError: false
    })

    render(<Character />)

    const loader = screen.getByTestId('skeleton')

    expect(loader).not.toBeNull()
  })

  it('Debe mostrar el personaje si se ha cargado correctamente', () => {


 
    mockUseParams.mockReturnValue({ id: '1' })
    mockUseQuery.mockReturnValue({
      isLoading: false,
      data: mockCharacterData,
      isError: false
    })

 

    const { container } = render(<Character />);
    

    expect(container).toMatchSnapshot()

    const name = screen.getByText('John Doe')
    const description = screen.getByText('Test description')
    const image = screen.getByAltText('John Doe')

    expect(name).not.toBeNull()
    expect(description).not.toBeNull()
    expect(image).not.toBeNull()
    expect(image.getAttribute('src')).toBe('https://example.com/image.jpg')




  })

  it("Debe renderizar el commic correctamente ", () => {


    mockUseParams.mockReturnValue({ id: '1' })
    mockUseQuery.mockReturnValue({
      isLoading: false,
      data: mockCommicsData,
      isError: false
    })
    const { container } = render(<Character />);
    

    expect(container).toMatchSnapshot()

    expect(screen.findByText('Amazing Spider-Man')).not.toBeNull()
    expect(screen.getByAltText('Amazing Spider-Man')).not.toBeNull()
    expect(screen.getByAltText('Amazing Spider-Man').getAttribute('src')).toBe('https://example.com/commic-image.jpg')

  })
})
