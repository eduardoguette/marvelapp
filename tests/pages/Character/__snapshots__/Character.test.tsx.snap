// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Character Component > Debe mostrar el personaje si se ha cargado correctamente 1`] = `
<div>
  <div>
    <section
      class="_characterHero_ad5e15"
    >
      <article
        class="_characterCard_ad5e15"
        style="opacity: 0; transform: translateX(-50px) translateZ(0);"
      >
        <img
          alt="John Doe"
          class="_characterImage_ad5e15"
          src="https://example.com/image.jpg"
        />
        <div
          class="_characterInfo_ad5e15"
        >
          <header
            class="_characterCardHeader_ad5e15"
          >
            <h2
              class="_characterTitle_ad5e15"
              style="opacity: 0;"
            >
              John Doe
            </h2>
            <button
              class="_button_6bd7a6"
              data-testid="action-component"
            >
              <svg
                fill="none"
                height="24"
                viewBox="0 0 26 25"
                width="24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M7 1.72192L7.51885 0.867058L7 0.55215L6.48115 0.867058L7 1.72192ZM13 5.36354L12.4812 6.21841L13 6.53332L13.5188 6.21841L13 5.36354ZM1 5.36354L0.481151 4.50868L0 4.8007V5.36354H1ZM1 13.167H0V13.6285L0.351203 13.928L1 13.167ZM13 23.3982L12.3512 24.1592L13 24.7123L13.6488 24.1592L13 23.3982ZM25 13.167L25.6488 13.928L26 13.6285V13.167H25ZM25 5.36354H26V4.8007L25.5188 4.50868L25 5.36354ZM19 1.72192L19.5189 0.867058L19 0.55215L18.4812 0.867058L19 1.72192ZM6.48115 2.57679L12.4812 6.21841L13.5188 4.50868L7.51885 0.867058L6.48115 2.57679ZM1.51885 6.21841L7.51885 2.57679L6.48115 0.867058L0.481151 4.50868L1.51885 6.21841ZM2 13.167V5.36354H0V13.167H2ZM13.6488 22.6373L1.6488 12.406L0.351203 13.928L12.3512 24.1592L13.6488 22.6373ZM13.6488 24.1592L25.6488 13.928L24.3512 12.406L12.3512 22.6373L13.6488 24.1592ZM26 13.167V5.36354H24V13.167H26ZM25.5188 4.50868L19.5189 0.867058L18.4812 2.57679L24.4812 6.21841L25.5188 4.50868ZM18.4812 0.867058L12.4812 4.50868L13.5188 6.21841L19.5189 2.57679L18.4812 0.867058Z"
                  fill="currentColor"
                />
              </svg>
            </button>
          </header>
          <p
            style="opacity: 0;"
          >
            Test description
          </p>
        </div>
      </article>
    </section>
    <section
      class="_characterContainer_ad5e15"
    >
      <article
        class="_characterCommics_ad5e15"
      >
        <h2
          class="_characterTitle_ad5e15"
        >
          Commics
        </h2>
        <ul
          class="_carrouselList_a584d8"
        >
          <div
            style="opacity: 0;"
          >
            <li
              class="_carrouselItem_a584d8"
            >
              <article>
                <img
                  class="_commicImage_5686fd"
                  height="252.79"
                  src="https://example.com/image.jpg"
                  width="168.53"
                />
                <p
                  class="_commicTitle_5686fd"
                />
                <span
                  class="_commicCaptionDate_5686fd"
                />
              </article>
            </li>
          </div>
        </ul>
      </article>
    </section>
  </div>
</div>
`;

exports[`Character Component > Debe renderizar el commic correctamente  1`] = `
<div>
  <div>
    <section
      class="_characterHero_ad5e15"
    >
      <article
        class="_characterCard_ad5e15"
        style="opacity: 0; transform: translateX(-50px) translateZ(0);"
      >
        <img
          class="_characterImage_ad5e15"
          src="https://example.com/commic-image.jpg"
        />
        <div
          class="_characterInfo_ad5e15"
        >
          <header
            class="_characterCardHeader_ad5e15"
          >
            <h2
              class="_characterTitle_ad5e15"
              style="opacity: 0;"
            />
            <button
              class="_button_6bd7a6"
              data-testid="action-component"
            >
              <svg
                fill="none"
                height="24"
                viewBox="0 0 26 25"
                width="24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M7 1.72192L7.51885 0.867058L7 0.55215L6.48115 0.867058L7 1.72192ZM13 5.36354L12.4812 6.21841L13 6.53332L13.5188 6.21841L13 5.36354ZM1 5.36354L0.481151 4.50868L0 4.8007V5.36354H1ZM1 13.167H0V13.6285L0.351203 13.928L1 13.167ZM13 23.3982L12.3512 24.1592L13 24.7123L13.6488 24.1592L13 23.3982ZM25 13.167L25.6488 13.928L26 13.6285V13.167H25ZM25 5.36354H26V4.8007L25.5188 4.50868L25 5.36354ZM19 1.72192L19.5189 0.867058L19 0.55215L18.4812 0.867058L19 1.72192ZM6.48115 2.57679L12.4812 6.21841L13.5188 4.50868L7.51885 0.867058L6.48115 2.57679ZM1.51885 6.21841L7.51885 2.57679L6.48115 0.867058L0.481151 4.50868L1.51885 6.21841ZM2 13.167V5.36354H0V13.167H2ZM13.6488 22.6373L1.6488 12.406L0.351203 13.928L12.3512 24.1592L13.6488 22.6373ZM13.6488 24.1592L25.6488 13.928L24.3512 12.406L12.3512 22.6373L13.6488 24.1592ZM26 13.167V5.36354H24V13.167H26ZM25.5188 4.50868L19.5189 0.867058L18.4812 2.57679L24.4812 6.21841L25.5188 4.50868ZM18.4812 0.867058L12.4812 4.50868L13.5188 6.21841L19.5189 2.57679L18.4812 0.867058Z"
                  fill="currentColor"
                />
              </svg>
            </button>
          </header>
          <p
            style="opacity: 0;"
          />
        </div>
      </article>
    </section>
    <section
      class="_characterContainer_ad5e15"
    >
      <article
        class="_characterCommics_ad5e15"
      >
        <h2
          class="_characterTitle_ad5e15"
        >
          Commics
        </h2>
        <ul
          class="_carrouselList_a584d8"
        >
          <div
            style="opacity: 0;"
          >
            <li
              class="_carrouselItem_a584d8"
            >
              <article>
                <img
                  alt="Amazing Spider-Man"
                  class="_commicImage_5686fd"
                  height="252.79"
                  src="https://example.com/commic-image.jpg"
                  width="168.53"
                />
                <p
                  class="_commicTitle_5686fd"
                >
                  Amazing Spider-Man
                </p>
                <span
                  class="_commicCaptionDate_5686fd"
                >
                  2021
                </span>
              </article>
            </li>
          </div>
        </ul>
      </article>
    </section>
  </div>
</div>
`;
